src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -159,40 +166,53 @@ private Task WriteBadRequestResponseAsync(HttpContext context, IDocumentWriter w
+        private async Task WriteResponseAsync<TResult>(HttpContext context, IDocumentWriter writer, TResult result)
         {
             context.Response.ContentType = "application/json";
             context.Response.StatusCode = 200; // OK
 
-            return writer.WriteAsync(context.Response.Body, result);
+            await writer.WriteAsync(context.Response.Body, result);
         }
-
-        private bool Deserialize(Stream stream, out GraphQLRequest single, out GraphQLRequest[] batch)
+       
+        private async Task<(bool Success, GraphQLRequest Request, GraphQLRequest[] Requests, Exception)> Deserialize(Stream stream)
         {
-            single = null;
-            batch = null;
 
             // Do not explicitly or implicitly (via using, etc.) call dispose because StreamReader will dispose inner stream.
             // This leads to the inability to use the stream further by other consumers/middlewares of the request processing
             // pipeline. In fact, it is absolutely not dangerous not to dispose StreamReader as it does not perform any useful
             // work except for the disposing inner stream.
             var reader = new StreamReader(stream);
-            using (var jsonReader = new JsonTextReader(reader) { CloseInput = false })
+
+            // There is no async peek, so need to check the type of jToken returned
+            // If anything other than { or ] or some other error then return false
+            try
             {
-                switch (reader.Peek())
+                using (var jsonReader = new JsonTextReader(reader) {CloseInput = false})
                 {
-                    case '{':
-                        single = _serializer.Deserialize<GraphQLRequest>(jsonReader);
-                        return true;
+                    var jToken = await JToken.LoadAsync(jsonReader).ConfigureAwait(false);
 
-                    case '[':
-                        batch = _serializer.Deserialize<GraphQLRequest[]>(jsonReader);
-                        return true;
+                    switch (jToken)
+                    {
+                        case JObject _:
+                        {
+                            var single = jToken.ToObject<GraphQLRequest>();
+                            return (true, single, null, null);
+                        }
+                        case JArray _:
+                        {
+                            var batch = jToken.ToObject<GraphQLRequest[]>();
+                            return (true, null, batch, null);
+                        }
 
-                    default:
-                        return false; // fast return with BadRequest without reading request stream
+                        default:
+                            return (false, null, null, null);
+                    }
                 }
             }
+            catch (Exception exception)
+            {
+                return (false, null, null, exception); 
```

Comment:
You are swallowing an exception. I don’t see that it is used somehow. The current implementation allows the exception to flow up the call stack and be processed by the corresponding http request middleware. I don't think swallowing an exception is the right way.

Reply:
I agree with you here.  It was because I could not use the Peek() method asynchronously so relied on the exception instead.   However, as you said all exceptions will get consumed.  Let's abandon this request, and will look at the code again.

Reply:
Will you close this PR?

Reply:
Yep will do.  By the way, just spent some time going through AspNetCore source code to see how the framework handles this as I was puzzled why this error was not coming up all the time.  Throughout the code it gets around this by taking the incoming HttpRequestStream and asynchronously reading it into FileBufferringReadStream which then gets used with the json deserialiser.

So if we apply the same pattern the code change to fix this will be quite minimal.